### PR TITLE
Factor out auth and forms. Fixes #12950

### LINKED ISSUE #12950: Factor out Forms and Authentication from Components
This came out of API review.

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
This adds two new packages, both of which are part of the shared framework *and* ship as NuGet packages for client-side Blazor:

 * `Microsoft.AspNetCore.Components.Authorization`
 * `Microsoft.AspNetCore.Components.Forms`

### Questions

**For the authorization classes, should we use the namespace `Microsoft.AspNetCore.Components.Authorization`?**
Currently this retains their existing namespaces, so they are all in `M.A.Components`. This differs from all the forms classes, which already have the namespaces `M.A.Components.Forms`. Since the separation of packages is a user-visible feature (in client-side Blazor at least, since there's no shared framework), it now feels more justified to emphasise the seam by having a distinct namespace for `.Authorization`.

**Should .Web reference .Authorization?**
Initially that's what I did, as it's what we discussed before. But I think this change achieves more if we *don't* have that reference. I'm not sure why we'd want `.Web` to be an uber-package that bundles all the things, especially considering that client-side Blazor projects have to reference `.Web` and you want to minimize unused dependencies (even when using the linker). This PR currently does *not* have a reference from .Web to .Authorization, and it works perfectly well.

**Should .Web reference .Forms?**
Currently in this PR it does, because .Web contains `InputText` etc. that depend on Forms APIs such as `EditContext`. If we want to take this a step further, we could also create a `M.A.Components.Forms.Web` package so that `.Web` itself doesn't have to reference `.Forms`. This means that transitively, default client-side Blazor projects wouldn't have to reference `System.ComponentModel.Annotations`. I haven't gone that far because I think this warrants more discussion about how far is too far when it comes to decoupling packages. Even if we do this, we might still want the default client-side template to reference `.Forms.Web` otherwise it's hard to discover how to do data editing, which is such a core scenario.

### CODE REVIEW

File: src/Components/Blazor/Blazor/src/Hosting/WebAssemblyHostBuilder.cs
Comment (SteveSandersonMS) [inline src/Components/Blazor/Blazor/src/Hosting/WebAssemblyHostBuilder.cs]: It's client-side Blazor so I'm not dealing with it now, but if we do this, we could drop the `M.E.Options` dependency from default client-side Blazor projects.

File: src/Mvc/Mvc.ViewFeatures/src/Microsoft.AspNetCore.Mvc.ViewFeatures.csproj
Comment (SteveSandersonMS) [inline src/Mvc/Mvc.ViewFeatures/src/Microsoft.AspNetCore.Mvc.ViewFeatures.csproj]: Needed for the code that sets up the authentication state during prerendering.

**Comment by SteveSandersonMS** [comment]
**Update** We discussed offline, and the answer to all the questions above is "keep it as it is in this PR".

**Comment by SteveSandersonMS** [comment]
Actually we do want to put Authorization classes in their own namespace.

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj
Comment (rynowak) [inline src/Components/Components/src/Microsoft.AspNetCore.Components.csproj]: Do we really want to undo this? Wouldn't this just get put back in the future if we decide to use a `.razor` file here?

**Review by rynowak — APPROVED** [review]

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj
Comment (SteveSandersonMS) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj]: Sure, I thought we’d just put it back if we have a reason to. It doesn’t produce any visible difference to the compiled package does it? As in, there’s nothing breaking or even visible about putting it back later.

I don’t mind. I’m happy to put it back if there’s a preference for that.

Comment (rynowak) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj]: I don't really take issue either way. It just seems like homework.

**Comment by JunTaoLuo** [comment]
Rebased on https://github.com/aspnet/AspNetCore/pull/12936 since there are conflicts.

File: eng/SharedFramework.Local.props
Comment (dougbu) [inline eng/SharedFramework.Local.props]: Curious: Why do these need to be in the shared framework?

Comment (rynowak) [reply eng/SharedFramework.Local.props]: This is a relayering of some of the parts of Blazor that are already in the shared framework 😆 

Server-side Blazor is a shared-framework feature.

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj
Comment (SteveSandersonMS) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj]: OK, since you don't mind I'll leave it as-is. We can trivially put it back if there's reason to later.

**Comment by SteveSandersonMS** [comment]
@JunTaoLuo How do you want to go about merging this? The components parts of this PR are done and ready to merge, but I don't know if the multitargeting change you rebased it onto is ready to go in.

If your parts of this are done, do you want to merge it?

**Comment by JunTaoLuo** [comment]
I rebased after merging the multitargeting PR, this is ready to go in.

**Comment by JunTaoLuo** [comment]
Hmm the test failure looks like it might be related to the change:
```
 Templates.Test.BlazorServerTemplateTest.BlazorServerTemplateWorks_IndividualAuth(useLocalDB: False)
Project new blazorserver  --auth Individual failed to publish.\r\nProcess exited with code 1\nStdErr: \nStdOut: Microsoft (R) Build Engine version 16.3.0-preview-19329-01+d31fdbf01 for .NET Core\r\nCopyright (C) Microsoft Corporation. All rights reserved.\r\n\r\n  Restore completed in 19.99 ms for F:\\workspace\\_work\\1\\s\\src\\ProjectTemplates\\test\\bin\\Release\\netcoreapp3.0\\TestTemplates\\AspNet.blazorserverindividual.ekm3gq2pvfa\\AspNet.blazorserverindividual.ekm3gq2pvfa.csproj.\r\n  You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview\r\nStartup.cs(40,32): error CS0246: The type or namespace name 'AuthenticationStateProvider' could not be found (are you missing a using directive or an assembly reference?) [F:\\workspace\\_work\\1\\s\\src\\ProjectTemplates\\test\\bin\\Release\\netcoreapp3.0\\TestTemplates\\AspNet.blazorserverindividual.ekm3gq2pvfa\\AspNet.blazorserverindividual.ekm3gq2pvfa.csproj]\r\nStartup.cs(40,22): error CS0311: The type 'AspNet.blazorserverindividual.ekm3gq2pvfa.Areas.Identity.RevalidatingIdentityAuthenticationStateProvider<Microsoft.AspNetCore.Identity.IdentityUser>' cannot be used as type parameter 'TImplementation' in the generic type or method 'ServiceCollectionServiceExtensions.AddScoped<TService, TImplementation>(IServiceCollection)'. There is no implicit reference conversion from 'AspNet.blazorserverindividual.ekm3gq2pvfa.Areas.Identity.RevalidatingIdentityAuthenticationStateProvider<Microsoft.AspNetCore.Identity.IdentityUser>' to 'AuthenticationStateProvider'. [F:\\workspace\\_work\\1\\s\\src\\ProjectTemplates\\test\\bin\\Release\\netcoreapp3.0\\TestTemplates\\AspNet.blazorserverindividual.ekm3gq2pvfa\\AspNet.blazorserverindividual.ekm3gq2pvfa.csproj]\r\n\r\nExpected: True\r\nActual:   False
   at Templates.Test.BlazorServerTemplateTest.BlazorServerTemplateWorks_IndividualAuth(Boolean useLocalDB) in /_/src/ProjectTemplates/test/BlazorServerTemplateTest.cs:line 100
--- End of stack trace from previous location where exception was thrown ---
Output:
==> F:\workspace\_work\1\s\.dotnet\dotnet.exe new blazorserver --debug:custom-hive "F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\TestTemplates\\Hives\f2ac4a6d-31e5-4642-bcbb-8bb14b4078bf\.templateengine" --auth Individual -o F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\TestTemplates\AspNet.blazorserverindividual.ekm3gq2pvfa [F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\]
The template "Blazor Server App" was created successfully.
This template contains technologies from parties other than Microsoft, see https://aka.ms/aspnetcore/3.0-third-party-notices for details.

Processing post-creation actions...
Running 'dotnet restore' on F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\TestTemplates\AspNet.blazorserverindividual.ekm3gq2pvfa\AspNet.blazorserverindividual.ekm3gq2pvfa.csproj...
  Restore completed in 218.71 ms for F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\TestTemplates\AspNet.blazorserverindividual.ekm3gq2pvfa\AspNet.blazorserverindividual.ekm3gq2pvfa.csproj.

Restore succeeded.

Publishing ASP.NET application...
==> F:\workspace\_work\1\s\.dotnet\dotnet.exe publish -c Release -p:PublishWithAspNetCoreTargetManifest=false [F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\TestTemplates\AspNet.blazorserverindividual.ekm3gq2pvfa]
Microsoft (R) Build Engine version 16.3.0-preview-19329-01+d31fdbf01 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 19.99 ms for F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\TestTemplates\AspNet.blazorserverindividual.ekm3gq2pvfa\AspNet.blazorserverindividual.ekm3gq2pvfa.csproj.
  You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview
Startup.cs(40,32): error CS0246: The type or namespace name 'AuthenticationStateProvider' could not be found (are you missing a using directive or an assembly reference?) [F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\TestTemplates\AspNet.blazorserverindividual.ekm3gq2pvfa\AspNet.blazorserverindividual.ekm3gq2pvfa.csproj]
Startup.cs(40,22): error CS0311: The type 'AspNet.blazorserverindividual.ekm3gq2pvfa.Areas.Identity.RevalidatingIdentityAuthenticationStateProvider<Microsoft.AspNetCore.Identity.IdentityUser>' cannot be used as type parameter 'TImplementation' in the generic type or method 'ServiceCollectionServiceExtensions.AddScoped<TService, TImplementation>(IServiceCollection)'. There is no implicit reference conversion from 'AspNet.blazorserverindividual.ekm3gq2pvfa.Areas.Identity.RevalidatingIdentityAuthenticationStateProvider<Microsoft.AspNetCore.Identity.IdentityUser>' to 'AuthenticationStateProvider'. [F:\workspace\_work\1\s\src\ProjectTemplates\test\bin\Release\netcoreapp3.0\TestTemplates\AspNet.blazorserverindividual.ekm3gq2pvfa\AspNet.blazorserverindividual.ekm3gq2pvfa.csproj]
```

Care to take a look @SteveSandersonMS

**Comment by SteveSandersonMS** [comment]
Yes, thanks. Pushed an update. Will check the results now.

**Pull request merged**
